### DISCUSSION TITLE
Resetting robot using code weird behaviour

### DISCUSSION THREAD

**Discussion by HeChengHui** [cyberbotics/webots#3838] [Q&A]
Hi, I am building a programme that moves the robot back to the starting point if it reaches some undesired place.
I do so by calling a reset function:
```
def reset(self):
        MotorFrontLeftW.setVelocity(0)
        MotorBackLeftW.setVelocity(0)
        MotorFrontRightW.setVelocity(0)
        MotorBackRightW.setVelocity(0)
        MotorFrontLeftW.setPosition(float('inf'))
        MotorBackLeftW.setPosition(float('inf'))
        MotorFrontRightW.setPosition(float('inf'))
        MotorBackRightW.setPosition(float('inf'))
        robot_translateion_field.setSFVec3f(self.STARTING_POS)
        robot_rotation_field.setSFRotation(self.STARTING_ROT)
        robot_node.resetPhysics()
```

However, it will sometimes just implode(?) or collapse when moved to starting position. 

https://user-images.githubusercontent.com/84503515/138908903-7dcf196d-104d-4c56-8e84-b42fa1eebebb.mp4

The Track node has no BoundingObject and only Floor node has a BoundingObject of a 1mx1m plane. Any idea on how not to let that happen?

**Comment by HeChengHui** [comment]
i think it may be something related to how the robot can somehow clip through bounding objects?
![image](https://user-images.githubusercontent.com/84503515/138964075-91bec6b3-3ffc-4cbe-ad06-c0b30b1ee668.png)

**Comment by ad-daniel** [comment]
Is `STARTING_POS` above ground or as you've shown in the second picture? Have you tried resetting slightly higher vertically?
If the starting position is already penetrating the ground it might indeed provoke these sort of explosions as the physics engine pushes the robot away (in this case too violently) in order to not break the constraint. 

Also, does the same occur if you decrease the timestep significantly? It could be that due to the weight of the robot, 1 timestep after reset if finds itself penetrating the ground.

**Comment by HeChengHui** [comment]
@ad-daniel Thank you for your feedback. 

I have tried setting translation such that the wheels are above and below the floor plane, and both cause the same reaction (setting above seems to have a much higher probability of it happening).  I have also thought that it might be the weight of the robot so I made all the nodes with `physics` to have `mass -1`. 
Example:
```
physics Physics {
      mass -1
      centerOfMass [ 0.007500 0.000000 -0.010000 ]
    }
```
But doing so doesn't seem to resolve the situation. I can upload the proto file if you would like to take a look.

I have not tried the timestep method so I would give it a try first.

**Reply by ad-daniel** [reply]
By setting mass to -1 it automatically computes the mass from the volume of the objects and the default density of `1000 kg/m^3`, which might not be appropriate. It might be better to set instead `density = -1` and attribute appropriate masses to the different parts

**Reply by HeChengHui** [reply]
I see. The robot initially has some mass to each bounding object but it also displayed the same behaviour.

In terms of setting a low timestep, the below video is when I decrease my timestep from 40 -> 5:

https://user-images.githubusercontent.com/84503515/139030034-5639b119-f0fa-435a-a81d-4b3d4e0fed3f.mp4

**Reply by ad-daniel** [reply]
It's odd indeed, could you provide the proto/world/controller ?

**Reply by HeChengHui** [reply]
[PROJECT.zip](https://github.com/cyberbotics/webots/files/7424616/PROJECT.zip)

for this project, I am trying to train this robot to complete the track using DQN. I have set it to reset everytime it goes too far from the track or goes to a certain region of the track (black regions) in the incorrect order or runs out of time (90s simulation time)
`ENV.reset() ` is the reset function.

Currently, to combat this clipping, I have implemented a wait function in `VectorMove.py` so that the robot will wait till the clipping is gone before taking an action.

**Reply by HeChengHui** [reply]
@ad-daniel 
I have made it a bit better by changing the starting rotation to when it is stabilised on ground level. Due to the back wheel being smaller, I believe this cause the robot to rock back and forth when released parallel to the ground. 

However, I am not sure why does it clip further to the ground as compared to other robots like the SurveyorSrv1, which is another track robot.
Before hitting ground plane:
![image](https://user-images.githubusercontent.com/84503515/139185136-666e378b-546d-4e20-bc81-c43284fb922b.png)
After hitting ground plane:
![image](https://user-images.githubusercontent.com/84503515/139185166-19875b7c-c631-4cad-b84c-5ded8e1fe2fd.png)

I have removed all mass nodes as discussed earlier so all physics nodes are empty.

**Reply by ad-daniel** [reply]
I've simplified the world/controller to the maximum leaving only the vehicle on a plane and honestly it seems quite stable (although 40 of basic timestep is excessive)

The controller only resets the position and then makes it move forward after a bit.

https://user-images.githubusercontent.com/44834743/139234039-74b1c86b-14d0-4f16-b41b-93bb5cbe3a9a.mp4

That tells me that the modeling of the robot itself is correct, and even  the resetting appears to be happening correctly so at this point I'm more tempted to believe the issue might either be by how you defined the ground or by the controller.

[PROJECT_v2.zip](https://github.com/cyberbotics/webots/files/7433173/PROJECT_v2.zip)

**Reply by HeChengHui** [reply]
Thank you for your help!

I will give it a try and change my environment. 
I am using a timestep of 40 to get 25fps on the camera.

**Reply by ad-daniel** [reply]
Controller and world can have different timesteps (as long as they are multiples of each other). So you can have a low timestep (in `worldInfo`), and the while loop in your controller (`while robot.step(timestep) != -1`) you can set a different value for it.

**Reply by HeChengHui** [reply]
Hmm I see.  
But is there any benefits from using a small global timestep?
And why would it be not advisable to use a timestep like 40?

**Reply by ad-daniel** [reply]
When the timestep is large, it's easier for the wheels to break the constraints of the ground (i.e it sinks into it), so although part of this penetration gets corrected the following timesteps, when the next clock tick comes, it will happen again and again. This isn't ideal, first because it's not very realistic and because it can destabilize the system as it attempts to correct the penetration over and over (especially if you don't make the contact properties between the wheel and the ground softer (increasing softCFM and softERP). If they are hard, it will react more violently to these violations). By lowering the timestep, you're diminishing this phenomena. Of course it also makes the simulation slower, so you'll need to balance the two things. I would drop it at least to 10 or 20.

**Reply by HeChengHui** [reply]
Thank you for the explanation, I will try out a timestep of 20 next.
